sp-sim/src/server.rs

```diff
@@ -32,30 +33,48 @@ pub(crate) struct UdpServer {
 
 impl UdpServer {
     pub(crate) async fn new(
-        bind_address: SocketAddrV6,
-        multicast_addr: Option<Ipv6Addr>,
+        network_config: &NetworkConfig,
         log: &Logger,
     ) -> Result<Self> {
-        let sock =
-            Arc::new(UdpSocket::bind(bind_address).await.with_context(
-                || format!("failed to bind to {}", bind_address),
-            )?);
-
-        // If we don't have a multicast address, use a non-multicast address;
-        // this avoids some unslightly if/else blocks around log statements
-        // below without affecting logic (as we also have to handle
-        // non-multicast multicast addresses for CI below).
-        let multicast_addr = multicast_addr.unwrap_or(Ipv6Addr::LOCALHOST);
-
-        // In some environments where sp-sim runs (e.g., some CI runners), we're
-        // not able to join ipv6 multicast groups. In those cases, we're
-        // configured with a "multicast_addr" that isn't actually multicast, so
-        // don't try to join the group if we have such an address.
-        if multicast_addr.is_multicast() {
-            sock.join_multicast_v6(&multicast_addr, 0).with_context(|| {
-                format!("failed to join multicast group {}", multicast_addr)
-            })?;
-        }
+        let sock = match network_config {
+            // In some environments where sp-sim runs (e.g., some CI runners),
+            // we're not able to join ipv6 multicast groups. In those cases,
+            // we're configured with an address that isn't actually multicast,
+            // so don't try to join the group if we have such an address.
+            NetworkConfig::Simulated { bind_addr } => {
+                UdpSocket::bind(bind_addr).await.with_context(|| {
+                    format!("failed to bind to {}", bind_addr)
+                })?
+            }
+
+            NetworkConfig::Real {
+                bind_addr,
+                multicast_addr,
+                multicast_interface,
+            } => {
+                if !multicast_addr.is_multicast() {
+                    bail!("{multicast_addr} is not multicast!");
+                }
+
+                let interface_index =
+                    if_nametoindex(multicast_interface.as_str())?;
```

Comment:
Can we add context with the interface name we tried to look up to this error? As written if the interface name doesn't exist, I get

```
sp-sim: ENODEV: No such device
```

Reply:
70f76805e